docs/man/plan.1.md
---
title: plan
section: 1
footer: plan VERSION_PLACEHOLDER
header: plan User's Manual
author: Daniel Schauenberg <d@unwiredcouch.com>
date: DATE_PLACEHOLDER
---

<!-- This is the plan(1) man page, written in Markdown. -->
<!-- To generate the roff version, run `make man` -->

# NAME

plan — a command line tool for personal planning

# SYNOPSIS

`plan <subcommand> [<options>]`

# SUBCOMMANDS

`plan add-note <text> ...`
: Add a note entry to the current daily note. This will create the file if it doesn't exist already.

`plan add-todo <text> ...`
: Add a todo entry to the current daily note. This will create the file if it doesn't exist already.

`plan daily-prep`
: Create a daily note based on the configure template.

`plan get-assigned-issues`
: Get issues assigned to oneself and print them as a markdown checklist.

`plan import-screenshots --config_file=<plan.yaml> [--no-dry-run]`
: Import screenshots from a location into the plan folder.

`plan validate-config --config_file=<plan.yaml>`
: Validate a provided config file.


# DESCRIPTION

`plan` is a command line utility to help with personal planning and note
taking. The premise is that notes are plain text files (assumed to be in
markdown format) in a folder. The tooling can then generate a daily note based
on a configured (or default) template. The main interaction for notes is
intended to be your favourite editor while `plan` provides tooling around it.

# CONFIGURATION FILE

`plan` reads a `.plan.yaml` configuration file in the current directory. That
way the config file can be colocated in the notes directory and different
settings can be used for different notes directories.

An example configuration file in yaml format can look like this

```
recurring_tasks:
  Monday:
    - plan the week

  Friday:
    - update achievements doc

daily_template: |-
  ## Misc

  It's {{ .Weekday }} today.
  You have {{ len .RecurringTasks }} recurring tasks today:
  {{ range .RecurringTasks }}
  - [ ] {{- .Name }}
  {{- end }}

  You have {{ len .AssignedTasks }} issues assigned:
  {{- range .AssignedTasks }}
  - [ ] {{ .Name }} [link]({{ .URL }})
  {{- end }}

  ## Log

date_format: "2006-01-02"

github:
  task_query: "assignee:mrtazz org:github state:open"

screenshot_import:
  source: "~/Desktop"
  destination: "dailies/20060102_attachments"
  file_format: "Screenshot 2006-01-02 at 15.04.05.png"
```

Validity of a configuration file can be checked with the `validate-config`
subcommand.

## RECURRING TASKS

`plan` supports simple weekly recurring tasks which can be added to the
configuration file. This is an array of strings that are available in the
template via the `{{ .RecurringTasks }}` templating macro.

`plan` also provides assigned tasks via the `{{ .AssignedTasks }}` macro in
the same way. Tasks here are taken from GitHub and have a `{{  .Name }}` and
`{{ .URL }}` attribute available for template rendering. In order for this to
work, a task query in GitHub search format needs to be configured.

## NOTE TEMPLATE RENDERING

The daily note is rendered with a small set of context variables that can be
used in templating. The templating engine is Go's `text/template` and should
be expected to behave in the same way:

- `{{ .Weekday }}`: Prints the current day of the week
- `{{ .FormattedDate }}`: Prints the current date formatted based on the
  `date_format` config key (default: "2006-01-02")
- `{[ .RecurringTasks }}`: An array of recurring tasks for the day as set in the
  configuration file. Can be iterated over and accessed with `{{ .Name }}`
  within
- `{{ .AssignedTasks }}`: An array of assigned issues from GitHub, based on
  the configured `github.task_query`. Can be iterated over and accessed with
  `{{ .Name }}` and `{{ .URL }}`

# IMPORTING SCREENSHOTS

The `plan` subcommand `import-screenshots` provides a way to consolidate
screenshots in the plan folder for easier referencing in notes. On running the
command it will list all screenshots (this technically works with any file and
not just images) in the source directory that match the provided file format.
If the command is run with the `--no-dry-run` flag provided, the files will be
moved to the destination directory.

Both the file format as well as the destination directory can contain golang
time format directives which will be interpreted as such. That makes it
possible to extract dates from screenshots and file them in date base folders.
The config file section that is used for configuration is `screenshot_import`
and the section `CONFIGURATION FILE` above has an example.


# META OPTIONS AND COMMANDS

`--help`
: Show list of command-line options.

`version`
: Show version of plan.



# AUTHOR

plan is maintained by mrtazz.

**Source code:** `https://github.com/mrtazz/plan`

# REPORTING BUGS

- https://github.com/mrtazz/plan/issues

# SEE ALSO

- https://github.com/mrtazz/vim-plan


plan.go
package main

import (
	"fmt"
	"os"
	"strings"
	"time"

	"github.com/alecthomas/kong"
	"github.com/mrtazz/plan/pkg/github"
	log "github.com/sirupsen/logrus"

	"github.com/mrtazz/plan/pkg/config"
	"github.com/mrtazz/plan/pkg/dailies"
	"github.com/mrtazz/plan/pkg/screenshots"
)

var (
	flags struct {
		Debug      bool   `help:"Enable debug mode."`
		ConfigFile string `default:".plan.yaml" help:"path to the config file."`
		AddTodo    struct {
			Text []string `arg:"" name:"text" help:"text to add as todo."`
		} `cmd:"" help:"Add a todo to the current daily note."`
		AddNote struct {
			Text []string `arg:"" name:"text" help:"text to add as a note."`
		} `cmd:"" help:"Add a note to the current daily note."`
		DailyPrep struct {
			Day      string `help:"day to write daily note for in '2022-12-31' format"`
			NoDryRun bool   `help:"whether to actually write the daily note"`
		} `cmd:"" help:"create the daily note file"`
		GetAssignedIssues struct {
			Format string `default:"markdown" help:"which format to use for outputting issues"`
		} `cmd:"" help:"Retrieve assigned issues"`
		ImportScreenshots struct {
			NoDryRun bool `help:"whether to actually create folders and move files"`
		} `cmd:"" help:"Import screenshots to the plan folder"`
		ValidateConfig struct {
		} `cmd:"" help:"Validate the passed config and return."`
		Version struct {
		} `cmd:"" help:"print version and exit."`
	}

	version   = "dev"
	goversion = "na"
)

const (
	dayFlagFormat = "2006-01-02"
)

func main() {

	ctx := kong.Parse(&flags, kong.UsageOnError())
	switch ctx.Command() {

	case "add-todo <text>":
		todayPlan := dailies.NewNote()
		text := fmt.Sprintf("\n- [ ] %s\n", strings.Join(flags.AddTodo.Text, " "))
		appendToFile(todayPlan.Filepath(), text)

	case "add-note <text>":
		todayPlan := dailies.NewNote()
		text := fmt.Sprintf("\n- %s\n", strings.Join(flags.AddNote.Text, " "))
		appendToFile(todayPlan.Filepath(), text)

	case "daily-prep":
		dailyPrep()

	case "get-assigned-issues":
		getAssignedIssues()

	case "validate-config":
		if err := config.ValidateConfig(flags.ConfigFile); err != nil {
			fmt.Printf("Failed to validate config '%s', got error: %s\n", flags.ConfigFile, err.Error())
			os.Exit(1)
		}
		fmt.Printf("Config file '%s' is valid.\n", flags.ConfigFile)
		return

	case "import-screenshots":
		cfg, err := config.LoadConfigFromFile(flags.ConfigFile)
		if err != nil {
			log.WithFields(log.Fields{
				"error":       err.Error(),
				"config-file": flags.ConfigFile,
			}).Error("failed to parse config")
			os.Exit(1)
		}
		i := screenshots.NewImporter(
			cfg.ScreenshotImport.Source,
			cfg.ScreenshotImport.Destination,
			cfg.ScreenshotImport.FileFormat,
			flags.ImportScreenshots.NoDryRun,
		)

		if err = i.ImportToPlanFolder(); err != nil {
			fmt.Printf("Failed to import screenshots: %s\n", err.Error())
			os.Exit(1)
		}

	case "version":
		fmt.Printf("plan: version %s %s", version, goversion)
		return

	default:
		ctx.FatalIfErrorf(fmt.Errorf("Unknown command: " + ctx.Command()))
	}
}

func dailyPrep() {

	cfg, err := config.LoadConfigFromFile(flags.ConfigFile)
	if err != nil {
		log.WithFields(log.Fields{
			"error":       err.Error(),
			"config-file": flags.ConfigFile,
		}).Error("failed to parse config")
		os.Exit(1)
	}

	token := os.Getenv("ISSUES_TOKEN_GITHUB")

	assignedIssues, err := github.GetAssignedTasks(token, cfg.GitHub.TaskQuery)

	if err != nil {
		fmt.Println("error getting assigned issues")
		fmt.Println(err)
	}

	todayPlan := dailies.NewNote().WithAssignedTasks(assignedIssues).
		WithRecurringTasks(cfg.GetRecurringTasks(time.Now().Weekday().String()))

	if cfg.DateFormat != "" {
		todayPlan = todayPlan.WithDateFormat(cfg.DateFormat)
	}

	if cfg.DailyTemplate != "" {
		todayPlan = todayPlan.WithTemplate(cfg.DailyTemplate)
	}

	if flags.DailyPrep.Day != "" {
		t, err := time.Parse(dayFlagFormat, flags.DailyPrep.Day)
		if err != nil {
			fmt.Println("error parsing provided date")
			fmt.Println(err)
			os.Exit(1)
		}
		todayPlan = todayPlan.WithDate(t)
	}

	dailyNoteString, err := todayPlan.Render()
	if err != nil {
		fmt.Println("error rendering daily note")
		fmt.Println(err)
		os.Exit(1)
	}

	if flags.DailyPrep.NoDryRun {
		fmt.Println(dailyNoteString)
		appendToFile(todayPlan.Filepath(), dailyNoteString)
	} else {
		fmt.Println(dailyNoteString)
	}

}

func getAssignedIssues() {
	cfg, err := config.LoadConfigFromFile(flags.ConfigFile)
	if err != nil {
		log.WithFields(log.Fields{
			"error":       err.Error(),
			"config-file": flags.ConfigFile,
		}).Error("failed to parse config")
		os.Exit(1)
	}

	token := os.Getenv("ISSUES_TOKEN_GITHUB")

	assignedIssues, err := github.GetAssignedTasks(token, cfg.GitHub.TaskQuery)

	if err != nil {
		fmt.Println("error getting assigned issues")
		fmt.Println(err)
	}

	switch flags.GetAssignedIssues.Format {
	case "markdown":
		for _, task := range assignedIssues {
			fmt.Printf("- [ ] [%s](%s)\n", task.Name(), task.URL())

		}

	default:
		fmt.Printf("Unknown format: '%s'\n", flags.GetAssignedIssues.Format)
	}

}

func appendToFile(filename, content string) error {
	f, err := os.OpenFile(filename, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		return err
	}
	defer f.Close()
	_, err = f.WriteString(content)
	return err
}
